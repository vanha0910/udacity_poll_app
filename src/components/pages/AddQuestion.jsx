import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { saveQuestion, resetState } from "../../redux/question/questionsSlice";
import {  setUsers} from "../../redux/auth/authenticationSlice";
import { useNavigate } from "react-router-dom";

import Input from "../common/Input";
import Button from "../common/Button";
import Loader from "../common/Loader";
import Header from "../common/Header";
import { PATH } from "../../constansts";

const AddQuestion = () => {
  const [optionOne, setOptionOne] = useState("");
  const [optionTwo, setOptionTwo] = useState("");
  const [hasSubmit, setSubmit] = useState(false)

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const {  isLoading,  isCreatedSuccess,  addedQuestion} = useSelector((state) => state.questions);
  const { authedUser } = useSelector((state) => state.authentication);

  const handleOptionOneChange = (e) => {
    setOptionOne(e.target.value);
  };

  const handleOptionTwoChange = (e) => {
    setOptionTwo(e.target.value);
  };

  const handleSubmit = (e) => {
    setSubmit(true)
    e.preventDefault();   

    dispatch(saveQuestion({
      optionOneText: optionOne,
      optionTwoText: optionTwo,
      author: authedUser &&  authedUser.id
    }));
  };

  useEffect(() => {
    if(isCreatedSuccess) {
      navigate(PATH.HOME)
      dispatch(setUsers({author: authedUser && authedUser.id, question: addedQuestion && addedQuestion.id}))
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isCreatedSuccess, addedQuestion])

  useEffect(() => {
    return () => dispatch(resetState())
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <>
      <Header />
      <div className="add-question-container">
        <h2>Would you rather</h2>
        <p>Create your own Poll</p>
        <div className="add-question-form">
          <form onSubmit={handleSubmit}>
            <Input
              label="First Option"
              type="text"
              id="optionOne"
              value={optionOne}
              onChange={handleOptionOneChange}
              required
              errorMessage="First option is required"
              data-testid="first-option"
              hasError={hasSubmit && !optionOne}

            />
            <Input
              label="Second Option"
              type="text"
              id="optionTwo"
              value={optionTwo}
              onChange={handleOptionTwoChange}
              required
              errorMessage="Second option is required"
              data-testid="second-option"
              hasError={hasSubmit && !optionTwo}

            />
            <div className="center-div">
            <Button type="submit"   data-testid="form-button">
              Create Question
            </Button>
            </div>
          </form>
        </div>
        <Loader open={isLoading} />
      </div>
    </>
  );
};

export default AddQuestion;
