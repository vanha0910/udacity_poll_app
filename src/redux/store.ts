import { combineReducers, configureStore } from '@reduxjs/toolkit'
import authenticationReducer from './auth/authenticationSlice'
import questionReducer from './question/questionsSlice'

const rootReducer = combineReducers({
  authentication: authenticationReducer,
  questions: questionReducer
})
export function setupStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState
  })
}
export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']