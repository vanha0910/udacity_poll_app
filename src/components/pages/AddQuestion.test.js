import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"; // To extend Jest matchers

import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import store from "../../redux";
import AddQuestion from "./AddQuestion"; // Assuming the file path to your component

describe("AddQuestion Component", () => {
  test("renders input fields and submit button", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <AddQuestion />
        </Provider>
      </MemoryRouter>
    );

    const firstOptionInput = screen.getByTestId("first-option");
    const secondOptionInput = screen.getByTestId("second-option");
    expect(firstOptionInput).toBeInTheDocument();
    expect(secondOptionInput).toBeInTheDocument();

    const submitButton = screen.getByTestId("form-button");
    expect(submitButton).toBeInTheDocument();
  });

  test("submits form with valid inputs", async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <AddQuestion />
        </Provider>
      </MemoryRouter>
    );
    const firstOptionInput = screen.getByTestId("first-option");
    const secondOptionInput = screen.getByTestId("second-option");
    fireEvent.change(firstOptionInput, { target: { value: "Option 1" } });
    fireEvent.change(secondOptionInput, { target: { value: "Option 2" } });

    const submitButton = screen.getByTestId("form-button");
    fireEvent.click(submitButton);

    expect(window.location.pathname).toEqual("/");
  });
});
