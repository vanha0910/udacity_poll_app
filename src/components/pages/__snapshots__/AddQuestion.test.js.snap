// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Question renders AddQuestion component correctly 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="nav-list"
      >
        <li
          class="nav-item  active"
        >
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/add"
          >
            New
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="user-info"
    >
      <button
        class="logout-btn"
      >
        Logout
      </button>
    </div>
  </header>
  <div
    class="add-question-container"
  >
    <h2>
      Would you rather
    </h2>
    <p>
      Create your own Poll
    </p>
    <div
      class="add-question-form"
    >
      <form>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionOne"
          >
            First Option
          </label>
          <input
            class="form-control "
            id="optionOne"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionTwo"
          >
            Second Option
          </label>
          <input
            class="form-control "
            id="optionTwo"
            type="text"
            value=""
          />
        </div>
        <div
          class="center-div"
        >
          <button
            class="btn btn-primary mt-2"
            disabled=""
            type="submit"
          >
            Create Question
          </button>
        </div>
      </form>
    </div>
    <div
      class="loading-container"
      style="display: none;"
    >
      <div
        class="loading"
      />
    </div>
  </div>
</div>
`;

exports[`renders AddQuestion component correctly 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="nav-list"
      >
        <li
          class="nav-item  active"
        >
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/add"
          >
            New
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="user-info"
    >
      <button
        class="logout-btn"
      >
        Logout
      </button>
    </div>
  </header>
  <div
    class="add-question-container"
  >
    <h2>
      Would you rather
    </h2>
    <p>
      Create your own Poll
    </p>
    <div
      class="add-question-form"
    >
      <form>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionOne"
          >
            First Option
          </label>
          <input
            class="form-control "
            id="optionOne"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionTwo"
          >
            Second Option
          </label>
          <input
            class="form-control "
            id="optionTwo"
            type="text"
            value=""
          />
        </div>
        <div
          class="center-div"
        >
          <button
            class="btn btn-primary mt-2"
            disabled=""
            type="submit"
          >
            Create Question
          </button>
        </div>
      </form>
    </div>
    <div
      class="loading-container"
      style="display: none;"
    >
      <div
        class="loading"
      />
    </div>
  </div>
</div>
`;

exports[`renders Header component correctly 1`] = `
<div>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="nav-list"
      >
        <li
          class="nav-item  active"
        >
          <a
            class="nav-link"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/leaderboard"
          >
            Leaderboard
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link"
            href="/add"
          >
            New
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="user-info"
    >
      <button
        class="logout-btn"
      >
        Logout
      </button>
    </div>
  </header>
  <div
    class="add-question-container"
  >
    <h2>
      Would you rather
    </h2>
    <p>
      Create your own Poll
    </p>
    <div
      class="add-question-form"
    >
      <form>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionOne"
          >
            First Option
          </label>
          <input
            class="form-control "
            id="optionOne"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group"
        >
          <label
            class="mb-1 label"
            for="optionTwo"
          >
            Second Option
          </label>
          <input
            class="form-control "
            id="optionTwo"
            type="text"
            value=""
          />
        </div>
        <div
          class="center-div"
        >
          <button
            class="btn btn-primary mt-2"
            disabled=""
            type="submit"
          >
            Create Question
          </button>
        </div>
      </form>
    </div>
    <div
      class="loading-container"
      style="display: none;"
    >
      <div
        class="loading"
      />
    </div>
  </div>
</div>
`;
